core/Pos/Binary/Class/Primitive.hs

```diff
@@ -123,13 +107,52 @@ deserializeOrFail bs0 =
         BSL.Empty           -> k Nothing      >>= supplyAllInput BSL.Empty
     supplyAllInput _ (CBOR.Read.Fail bs _ exn) = return (Left (exn, bs))
 
--- | Strict variant of 'deserializeOrFail'.
-deserializeOrFail'
-    :: Bi a
-    => BS.ByteString
-    -> Either (CBOR.Read.DeserialiseFailure, BS.ByteString)
-              (a, BS.ByteString)
-deserializeOrFail' = deserializeOrFail . BSL.fromStrict
+    -- | Deserialize a Haskell value from the external binary representation.
+    --
+    -- This allows /input/ data to be provided incrementally, rather than all in one
+    -- go. It also gives an explicit representation of deserialisation errors.
+    --
+    -- Note that the incremental behaviour is only for the input data, not the
+    -- output value: the final deserialized value is constructed and returned as a
+    -- whole, not incrementally.
+    deserializeIncremental :: Bi a => ST s (CBOR.Read.IDecode s a)
+    deserializeIncremental =
+        CBOR.Read.deserialiseIncremental (runReaderT decode dc)
+
+-- | Try to deserialize value or fail.
+--
+-- /Throws/: @'CBOR.Read.DeserialiseFailure'@ if the given external
+-- representation is invalid or does not correspond to a value of the
+-- expected type.
+deserializeThrow :: (Bi a) => LByteString -> a
+deserializeThrow =
+    either throw identity . bimap fst fst . deserializeOrFailRaw def
+
+decodeFullProcess ::
+       forall a. (Bi a)
+    => Either (CBOR.Read.DeserialiseFailure, BS.ByteString) (a, BS.ByteString)
+    -> Either Text a
+decodeFullProcess = \case
+    Right (x, leftover) -> case BS.null leftover of
+        True  -> pure x
+        False ->
+            let msg = "decodeFull failed for " <> label (Proxy @a) <>
+                      "! Leftover found: " <> show leftover
+            in Left $ fromString msg
+    Left  (e, _) ->
+        Left $ fromString $ "decodeFull failed for " <> label (Proxy @a) <>
+                            ": " <> show e
+
+-- | Deserialize a Haskell value from the external binary representation,
+-- failing if there are leftovers. In a nutshell, the `full` here implies
+-- the contract of this function is that what you feed as input needs to
+-- be consumed entirely. NoCheck version corresponds to passing @dcNocheck@
+-- to decoder parser.
+decodeFull, decodeFullNoCheck :: forall a. Bi a => BS.ByteString -> Either Text a
+decodeFull =
+    decodeFullProcess . deserializeOrFailRaw def . BSL.fromStrict
+decodeFullNoCheck =
+    decodeFullProcess . deserializeOrFailRaw (DecoderConfig False). BSL.fromStrict
```

Comment:
space before the dot \</nitpick>

Reply:
Also, shouldn't it be `DecoderConfig True`?

(Btw, maybe you should write `def {_dcNoCheck = True}` to make such bugs easier to catch.)